### ISSUE TITLE
Optionally specify file order by a fileorder.fsx (or fileorder.txt or fileorder.json) file

### ISSUE THREAD

**Issue by baronfel** [fsharp/fslang-suggestions#95]
**Submitted by Don Syme on 4/12/2016 12:00:00 AM**  
**20 votes on UserVoice prior to migration**  

With F# becoming more and more multi-editor and cross-platform, it is becoming increasingly difficult to teach all build/edit tools about F#'s file order. The F# community are currently struggling to "update" each new build/edit tool to understand that F# actually needs a file order.

Part of the problem is that there is no standard textual way to specify this file order except as command line arguments, and these are not stored in an editable form. There is no standard way to specify the F# file order. We need an (optional) solution to this problem that is closer to home and doesn't involve modifying build/edit tools.

This proposal is one of three alternatives to deal with this problem in the F# language/compiler itself.
The specific proposal covered by this UV entry is to allow the F# compiler to optionally take a special file, tentatively called fileorder.fsx, which specifies the file order E.g.

``fileorder.fsx``:
```fsharp
#load "Directory/a.fs"
#load "b.fs"
```

This would be given to the F# compiler as a command-line input and otherwise everything would work as it does today.
Rules
* File references can be given in any order on the command-line.
* If present, the file order is taken from fileorder.fsx
* Using fileorder.fsx is optional
* fileorder.fsx would be hand-authored by the user
* An error would be given if fileorder.fsx doesn’t list all the files in the compilation.

Questions:
* Would we use a fileorder.fsx or fileorder.txt
* We need a way to refer to things like “obj/Debug/pars.fs” that are the intermediate outputs of other actions. I suggest wildcards
```
  #load "obj/*/pars.fs"
```
Here the meaning of wildcards is “all the inputs that match” not “all the files on disk that match”.

Disadvantages:
* It’s at least as tasteless than “#load/#require”
* It requires manual editing

Advantages:
* Single place of reference for file order.
* Requires no changes to F# language, “it just works” with any editor
* We could augment the F# compiler to automatically write out a fileorder.fsx if one doesn’t exist, which would allow the user to gradually switch to this model.

 Related alternatives: 
* Allow all declarations to be mutually referential and the compiler takes files in any order 
* Keep a file order, but infer it from #load/#require declarations. 

**[Original UserVoice Submission](https://fslang.uservoice.com/forums/245727-f-language/suggestions/13394442)**
[Archived Uservoice Comments](../tree/master/archive/suggestion-13394442-optionally-specify-file-order-by-a-fileorder-fsx.md#comments)

**Comment by cartermp** [comment]
cc @brettfo - it may worth chiming in here.

**Comment by brettfo** [comment]
Maybe I'm missing something, but we already have project files (*.fsproj) that explicitly list the compilation order, why not just use that?  Visual Studio and Visual Studio Code already support this file.

**Comment by nelak** [comment]
I believe #309 is a better solution to the problem. By augmenting the #load directive syntax we can easily improve the xplat story. The fsharp compiler already supports compiling an .fsx file with #load directives.

**Comment by xperiandri** [comment]
Why compiler won't resolve dependencies itself then?
What will be the difference between `open` and `#load` in concept?
It think that current way is absolutely fine.

**Comment by sparebytes** [comment]
*tldr*: This limitation casts doubt and might hurt adoption of what seems to be a powerful and succinct language.

I'd really like to make this a high priority so I'll leave my 2 cents here. I suppose there are many others like me with a similar background. I maintain projects in other languages with thousands of source files. Cyclical dependencies are rarely an issue I deal with in those projects but it does happen from time to time. It's easily worth the trade-off to not have to specify the order in which files must be compiled.

I was very interested in using F# but it didn't take me long to find this limitation and now I'm scared to continue.

- How much time will I spend editing the fsproj file?
- It feels like a bad omen. The compiler isn't my friend in this case. What other pain points are lurking?
- I sometimes refactor a ton of files. Fixing the fsproj will distract me from the real goal.
- I sometimes generate multiple new source files from a template to get started on a new feature. That process won't be as smooth.
- If I introduce F# to my coworkers and said, "Oh now you gotta open the fsproj and add the file in the right spot." they would look at me funny.
- Anything that relies on the IDE to solve this feels like a hack.
- Anything that relies on special directives also feels like a hack. `open` declarations should be enough.

Maybe my concerns be overblown. Many F# developers seem to be getting along just fine without it.

**Comment by 7sharp9** [comment]
@sparebytes You sort of get used to the limitation, you don't often generate a bunch of files with F# so I doubt you would hit that issue.  When refactoring you may end up having to tweak file order though, I often get to a point during refactoring where the dependency suddenly appears then you have to start thinking about the file and type ordering and pondering adding a rec namespace or switching `type` to `and` on definitions etc.

**Comment by sparebytes** [comment]
*Maybe* it's possible to use `FSharp.Compiler.Service` to solve it:

- Take a list or glob of `*.fs` files
- Loop over each file and
  1. Parse the unchecked tree
  2. Yield all public/internal module declarations and the namespace they reside in
  3. Yield all references that aren't declared within a higher scope of the same file (we filter the references more later on)
- Concatenate a list of all the exposed modules from all the files
- Filter the references in each file that match with a module in another file. This is to ignore references to DLLs and external libraries (Very tricky and not robust)
- Build a dependency graph
- ...
- Profit

I know I overlooked something. EG, I'm not familiar with the `rec` keyword and whether it applies across files.

**Comment by cartermp** [comment]
FWIW I've always disagreed that file ordering is a feature, as it is sometimes evangelized due to keeping cycles out of your F# dependencies, with my rationale being that people organize files and folders differently than their dependencies. But the pain of ordering is quite ameliorated by using an IDE, which is a possibility on every desktop OS at this point.

[Mechanic](https://github.com/fsprojects/Mechanic) is a tool [also available through VSCode](https://github.com/ionide/ionide-vscode-mechanic) that will order files for you, but it is missing a few features that are up for grabs. IDE integration in VS, VS for Mac, and Rider are missing though. But each of these has the ability to order files in the IDE in a pretty convenient way.

Additionally, with VS if you just use `Add File` (rather than the other `Add Above`/`Add Below` context menu items), it will automatically add the file to the top of the current scope you're adding to (root of project, folder, subfolder), meaning there's no work to do in getting code to compile from the outset. Only when you're moving files around and changing your in-project dependency graph does ordering come into play if you're using VS.

**Comment by 7sharp9** [comment]
@sparebytes rec allows mutually dependent types within a single rec namespace or rec module.  F# tends to have longer files and less of them in general.

**Comment by davidglassborow** [comment]
@sparebytes you might find this discussion of cyclic dependencies relevant - https://fsharpforfunandprofit.com/posts/cyclic-dependencies/

**Comment by dsyme** [comment]
Closing as, if we do anything allowing further mutual-reference, we won't do it like this
